pkg/hubble/relay/observer.go

```diff
@@ -20,17 +20,209 @@ import (
+
+	go func() {
+		defer close(sortedFlows)
+
+	flowsLoop:
+		for {
+			select {
+			case flow, ok := <-flows:
+				if !ok {
+					break flowsLoop
+				}
+				if pq.Len() == qlen {
+					f := pq.Pop()
+					select {
+					case sortedFlows <- f:
+					case <-ctx.Done():
+						return
+					}
+				}
+				pq.Push(flow)
+			case <-time.After(bufferDrainTimeout): // make sure to drain the queue when no new flow responses are received
+				if f := pq.Pop(); f != nil {
+					select {
+					case sortedFlows <- f:
+					case <-ctx.Done():
+						return
+					}
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
+		// drain the queue
+		for f := pq.Pop(); f != nil; f = pq.Pop() {
+			select {
+			case sortedFlows <- f:
+			case <-ctx.Done():
+				return
+			}
+		}
+
+	}()
+
+	return sortedFlows
+}
+
+func nodeStatusError(err error, nodeNames ...string) *observerpb.GetFlowsResponse {
+	msg := err.Error()
+	if s, ok := status.FromError(err); ok && s.Code() == codes.Unknown {
+		msg = s.Message()
+	}
+
+	return &observerpb.GetFlowsResponse{
+		NodeName: nodeTypes.GetName(),
+		Time:     ptypes.TimestampNow(),
+		ResponseTypes: &observerpb.GetFlowsResponse_NodeStatus{
+			NodeStatus: &relaypb.NodeStatusEvent{
+				StateChange: relaypb.NodeState_NODE_ERROR,
+				NodeNames:   nodeNames,
+				Message:     msg,
+			},
+		},
+	}
+}
+
+func nodeStatusEvent(state relaypb.NodeState, nodeNames ...string) *observerpb.GetFlowsResponse {
+	return &observerpb.GetFlowsResponse{
+		NodeName: nodeTypes.GetName(),
+		Time:     ptypes.TimestampNow(),
+		ResponseTypes: &observerpb.GetFlowsResponse_NodeStatus{
+			NodeStatus: &relaypb.NodeStatusEvent{
+				StateChange: state,
+				NodeNames:   nodeNames,
+			},
+		},
+	}
+}
+
+func aggregateErrors(
+	ctx context.Context,
+	responses <-chan *observerpb.GetFlowsResponse,
+	errorAggregationWindow time.Duration,
+) <-chan *observerpb.GetFlowsResponse {
+	aggregated := make(chan *observerpb.GetFlowsResponse, cap(responses))
+
+	var flushPending <-chan time.Time
+	var pendingResponse *observerpb.GetFlowsResponse
+
+	go func() {
+		defer close(aggregated)
+	aggregateErrorsLoop:
+		for {
+			select {
+			case response, ok := <-responses:
+				if !ok {
+					// flush any pending response before exiting
+					if pendingResponse != nil {
+						select {
+						case aggregated <- pendingResponse:
+						case <-ctx.Done():
+						}
+					}
+					return
+				}
+
+				// any non-error responses are directly forwarded
+				current := response.GetNodeStatus()
+				if current.GetStateChange() != relaypb.NodeState_NODE_ERROR {
+					select {
+					case aggregated <- response:
+						continue aggregateErrorsLoop
+					case <-ctx.Done():
+						return
+					}
+				}
+
+				// either merge with pending or flush it
+				if pending := pendingResponse.GetNodeStatus(); pending != nil {
+					if current.GetMessage() == pending.GetMessage() {
+						pending.NodeNames = append(pending.NodeNames, current.NodeNames...)
```

Comment:
As I understand it, there are no node names de-dup logic, right?

Reply:
Yes. However, as it is implemented right now, a node can throw an error only once during a request. There cannot be any duplicate nodes to begin with.

Reply:
Ah, right because there is no retry logic yet.